### ISSUE TITLE
[@typescript-eslint/require-await] false positive on async generators without `await`

### ISSUE THREAD

**Issue by mysticatea** [typescript-eslint/typescript-eslint#3507]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
    "plugins": ["@typescript-eslint"],
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "ecmaVersion": 2021,
        "sourceType": "module",
        "project": "tsconfig.json",
    },
    "rules": {
        "@typescript-eslint/require-await": "error",
    },
}
```

```json
{
    "compilerOptions": {
        "strict": true
    },
    "include": ["test.ts"]
}
```

```TS
async function* concat<T>(first: Promise<T> | T, it: AsyncIterable<T> | Iterable<T>): AsyncIterableIterator<T> {
    yield first // `yield` unwraps `Promise` (thennable) in async generator function
    yield* it   // `yield*` iterates `AsyncIterable` (having `[Symbol.asyncIterator]()` method) in async generator function
}
```

**Expected Result**

No errors.

**Actual Result**

```
~\dev\sandbox> eslint test.ts

~\dev\sandbox\test.ts
  1:1  error  Async generator function 'concat' has no 'await' expression  @typescript-eslint/require-await

✖ 1 problem (1 error, 0 warnings)
```

**Additional Info**

#1691 is related but seems not enough.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.26.1` |
| `@typescript-eslint/parser`        | `4.26.1` |
| `TypeScript`                       | `4.3.2` |
| `ESLint`                           | `7.28.0` |
| `node`                             | `16.1.0` |

**Comment by bradzacher** [comment]
Generators themselves are rarely used - let alone async generators.
Likely why nobody has encountered this.
